File: user-service/src/utils/response.util.ts
import { Response } from 'express';

export const success = (res: Response, message = 'Success', data?: any) =>
  res.status(200).json({ success: true, message, ...(data !== undefined ? { data } : {}) });

export const created = (res: Response, message = 'Created', data?: any) =>
  res.status(201).json({ success: true, message, ...(data !== undefined ? { data } : {}) });

export const updated = (res: Response, message = 'Updated', data?: any) =>
  res.status(200).json({ success: true, message, ...(data !== undefined ? { data } : {}) });

export const deleted = (res: Response, message = 'Deleted') =>
  res.status(200).json({ success: true, message });

export const noContent = (res: Response) => res.status(204).send();

export const error = (res: Response, message = 'Bad Request', statusCode = 400, err?: any) =>
  res.status(statusCode).json({ success: false, message, ...(err ? { error: err } : {}) });

export const unauthorized = (res: Response, message = 'Unauthorized') =>
  res.status(401).json({ success: false, message });

export const forbidden = (res: Response, message = 'Forbidden') =>
  res.status(403).json({ success: false, message });

export const notFound = (res: Response, message = 'Not found') =>
  res.status(404).json({ success: false, message });

export const methodNotAllowed = (res: Response, message = 'Method Not Allowed') =>
  res.status(405).json({ success: false, message });

export const requestTimeout = (res: Response, message = 'Request Timeout') =>
  res.status(408).json({ success: false, message });

export const duplicate = (res: Response, message = 'Duplicate resource', err?: any) =>
  res.status(409).json({ success: false, message, ...(err ? { error: err } : {}) });

export const gone = (res: Response, message = 'Gone') =>
  res.status(410).json({ success: false, message });

export const unsupportedMediaType = (res: Response, message = 'Unsupported Media Type') =>
  res.status(415).json({ success: false, message });

export const unprocessable = (res: Response, message = 'Unprocessable Entity', err?: any) =>
  res.status(422).json({ success: false, message, ...(err ? { error: err } : {}) });

export const tooManyRequests = (res: Response, message = 'Too Many Requests') =>
  res.status(429).json({ success: false, message });

export const internalError = (res: Response, message = 'Internal server error', err?: any) =>
  res.status(500).json({ success: false, message, ...(err ? { error: err } : {}) });

export const notImplemented = (res: Response, message = 'Not Implemented') =>
  res.status(501).json({ success: false, message });

export const serviceUnavailable = (res: Response, message = 'Service Unavailable') =>
  res.status(503).json({ success: false, message });

File: apps/auth-service/src/main/java/com/idoc/auth/util/ResponseUtil.java
package com.idoc.auth.util;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Consumer;

import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import com.idoc.auth.dto.response.PageResponse;
import com.idoc.auth.dto.response.UserResponse;
import com.idoc.auth.dto.response.AuthenticationResponse;

public class ResponseUtil {

  private static Map<String, Object> baseBody(boolean success, String message) {
    Map<String, Object> body = new LinkedHashMap<>();
    body.put("success", success);
    body.put("message", message);
    return body;
  }

  public static ResponseEntity<Map<String, Object>> custom(HttpStatus status, boolean success, String message, Consumer<Map<String, Object>> customizer) {
    Map<String, Object> body = baseBody(success, message);
    if (customizer != null) customizer.accept(body);
    return ResponseEntity.status(status).body(body);
  }

  public static ResponseEntity<Map<String, Object>> authentication(String message, AuthenticationResponse auth, Object data) {
    return custom(HttpStatus.OK, true, message, body -> {
      body.put("token", auth.getToken());
      body.put("user", auth.getUser());
      if (data != null) body.put("data", data);
    });
  }

  public static ResponseEntity<Map<String, Object>> user(String message, UserResponse user) {
    return custom(HttpStatus.OK, true, message, body -> {
      body.put("user", user);
    });
  }

  public static <T> ResponseEntity<Map<String, Object>> paged(String message, Page<T> page) {
    return custom(HttpStatus.OK, true, message, body -> {
      body.put("data", page.getContent());
      body.put("pagination", new PageResponse(
        page.getTotalElements(),
        page.getSize(),
        page.getNumber(),
        page.getTotalPages()
      ));
    });
  }

  public static ResponseEntity<Map<String, Object>> success(String message, Object data) {
    return custom(HttpStatus.OK, true, message, body -> {
      if (data != null) body.put("data", data);
    });
  }

  public static ResponseEntity<Map<String, Object>> created(String message, Object data) {
    return custom(HttpStatus.CREATED, true, message, body -> {
      if (data != null) body.put("data", data);
    });
  }

  public static ResponseEntity<Map<String, Object>> updated(String message, Object data) {
    return success(message, data);
  }

  public static ResponseEntity<Map<String, Object>> deleted(String message) {
    return success(message, null);
  }

  public static ResponseEntity<Void> noContent() {
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

  public static ResponseEntity<Map<String, Object>> error(String message, HttpStatus status, Object err) {
    return custom(status, false, message, body -> {
      if (err != null) body.put("error", err);
    });
  }

  // Các hàm error khác có thể gọi lại error() ở trên
  public static ResponseEntity<Map<String, Object>> unauthorized(String message) {
    return error(message, HttpStatus.UNAUTHORIZED, null);
  }

  public static ResponseEntity<Map<String, Object>> forbidden(String message) {
    return error(message, HttpStatus.FORBIDDEN, null);
  }

  public static ResponseEntity<Map<String, Object>> notFound(String message) {
    return error(message, HttpStatus.NOT_FOUND, null);
  }

  public static ResponseEntity<Map<String, Object>> methodNotAllowed(String message) {
    return error(message, HttpStatus.METHOD_NOT_ALLOWED, null);
  }

  public static ResponseEntity<Map<String, Object>> requestTimeout(String message) {
    return error(message, HttpStatus.REQUEST_TIMEOUT, null);
  }

  public static ResponseEntity<Map<String, Object>> duplicate(String message, Object err) {
    return error(message, HttpStatus.CONFLICT, err);
  }

  public static ResponseEntity<Map<String, Object>> gone(String message) {
    return error(message, HttpStatus.GONE, null);
  }

  public static ResponseEntity<Map<String, Object>> unsupportedMediaType(String message) {
    return error(message, HttpStatus.UNSUPPORTED_MEDIA_TYPE, null);
  }

  public static ResponseEntity<Map<String, Object>> unprocessable(String message, Object err) {
    return error(message, HttpStatus.UNPROCESSABLE_ENTITY, err);
  }

  public static ResponseEntity<Map<String, Object>> tooManyRequests(String message) {
    return error(message, HttpStatus.TOO_MANY_REQUESTS, null);
  }

  public static ResponseEntity<Map<String, Object>> internalError(String message, Object err) {
    return error(message, HttpStatus.INTERNAL_SERVER_ERROR, err);
  }

  public static ResponseEntity<Map<String, Object>> notImplemented(String message) {
    return error(message, HttpStatus.NOT_IMPLEMENTED, null);
  }

  public static ResponseEntity<Map<String, Object>> serviceUnavailable(String message) {
    return error(message, HttpStatus.SERVICE_UNAVAILABLE, null);
  }
}